### PR TITLE
Work with Multihashes directly

### PR DESCRIPTION
**Pull request by hsanjuan**
This will be breaking for anyone using filestore who has not migrated
the underlying storage and coverted CIDs to raw multihashes.

It keeps the outside interfaces intact except for `ListRes` which includes
a multihash rather than a Cid.

### CODE REVIEW

File: filestore.go
Comment (hsanjuan) [inline filestore.go:22]: Naming global variables `log` confuses my `goimports` and it's a pain for me personally, so I took the liberty.

File: util.go
Comment (hsanjuan) [inline util.go]: This is the only public-interface place affected. It is easy to return a raw cid v1 instead though. Worth causing a small breakage though, given the changes below.

File: filestore.go
Comment (Stebalien) [reply filestore.go:22]: Are you signing up to do this all across go-ipfs?

Comment (Stebalien) [reply filestore.go:22]: :pray:

File: fsrefstore.go
Comment (Stebalien) [inline fsrefstore.go:200]: note: I believe all filestore CIDs will be raw anyways.

File: go.mod
Comment (Stebalien) [inline go.mod]: 2.0

File: util.go
Comment (Stebalien) [reply util.go]: This will affect the filestore commands output. It is an experimental feature so I'm not really concerned but it's still something to keep in mind.

Comment (Stebalien) [reply util.go]: I guess we could turn them back to CIDs at that point.

Comment (Stebalien) [reply util.go]: Although... given that we _know_ that the type is Raw (all filestore blocks are ipld-raw), I'm not actually sure if this change is worth it.

File: util.go
Comment (Stebalien) [inline util.go]:
```suggestion
		mhash, keyErr := dshelp.DsKeyToMultihash(ds.RawKey(v.dsKey))
```

**Review by Stebalien — COMMENTED** [review]
Other than the nits, my only real concern is that switching _everything_ to multihashes is, strictly speaking, throwing away correct information that we have. We _only_ reference "raw leaves" with the filestore so we always know the codec.

File: filestore.go
Comment (hsanjuan) [reply filestore.go:22]: When I see it I certainly will update.

File: util.go
Comment (hsanjuan) [reply util.go]: So `FormatLong` method below did just that (turn back to CIDs). Will this change API output? I am now thinking that Multihashes do not know how to marshal themselves to json :/

File: util.go
Comment (hsanjuan) [reply util.go]: ack

**Comment by hsanjuan** [comment]
> switching everything to multihashes is, strictly speaking, throwing away correct information that we have

We are doing this in general right? IPFS will lose any clue on what type of content it is storing in the datastore, other than that provided by context (in this case knowing that /filestore things are raw). But that was the case with CID v0s anyways..

File: util.go
Comment (hsanjuan) [reply util.go]: I reverted and kept this as Cid.

**Comment by Stebalien** [comment]
> We are doing this in general right? IPFS will lose any clue on what type of content it is storing in the datastore, other than that provided by context (in this case knowing that /filestore things are raw). But that was the case with CID v0s anyways..

Not quite.

* The blockstore stores blocks with different versions/formats. We're intentionally throwing this information away so we don't have to re-download blocks when we're just changing the CID version.
* The filestore only stores CIDv1-raw blocks so we don't have this issue.

**Review by Stebalien — APPROVED** [review]
Now we just need the filestore migration.

**Comment by Stebalien** [comment]
Comment on where this lives in the migration process (context for anyone wondering):

This change isn't _strictly_ speaking necessary for stage 1 as (a) the blockstore interfaces still use CIDs and (b) all blocks in the filestore will be cidv1 raw blocks. However, when we migrate the blockstore interfaces to multihashes, we'd either have to convert inbound multihashes back into CIDs when looking them up in the filestore's section of the datastore _or_ we'd have to migrate at that point.

Migrating now and using multihashes everywhere is likely the cleanest solution.

**Comment by hsanjuan** [comment]
> 
> Migrating now and using multihashes everywhere is likely the cleanest solution.

I'll add this to the migration

**Comment by hsanjuan** [comment]
Do not merge: I need to set the right go-ipfs-blockstore dependency here before merging

**Comment by hsanjuan** [comment]
I have fixed the deps and merged. I tagged v1.0.0 even if this was not breaking and things worked without needing a migration because all filestore blocks are CidV1.Raw. Just in case someone, somewhere, has done something else with this or misused it somehow.

**Pull request merged**
